File: day13/day13a.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

int main() {
  // Read input from file
  std::ifstream file("day13.txt");
  std::string line;
  int start;

  // Parse first line
  std::getline(file, line);
  std::stringstream start_ss(line);
  start_ss >> start;

  // Parse second line, where bus numbers are comma-separated
  std::getline(file, line);
  std::stringstream buses_ss(line);
  std::string bus_str;
  int bus;

  int min_diff = start;
  int min_bus = -1;

  while(std::getline(buses_ss, bus_str, ',')) {
    // Extract the next bus number, if it's not 'x'
    if(bus_str != std::string("x")) {
      std::stringstream bus_str_ss(bus_str);
      bus_str_ss >> bus;

      int diff = (bus - (start % bus)) % bus;
      if(min_diff > diff) {
        min_diff = diff;
        min_bus = bus;
      }
    }
  }

  std::cout << min_bus * min_diff << std::endl;
  return 0;
}

File: day13/day13a.py
#!/usr/bin/python3

with open('day13.txt', 'r') as f:
    lines = f.readlines()
    start = int(lines[0])
    buses = [int(elem) for elem in lines[1].split(',') if elem != 'x']
    min_diff = start
    min_bus = -1
    for bus in buses:
        # Search through each bus, computing the difference from 'start' until the next multiple of bus
        diff = (bus - (start % bus)) % bus
        if min_diff > diff:
            min_diff = diff
            min_bus = bus
    print(min_bus * min_diff)

